### PR TITLE
[BC-376] Support snappy over gossip

### PR DESCRIPTION
**Pull request by mbaxter**
## PR Description
Add support for snappy compression over gossip.

Also:
* Fixed some ssz serialization issues:
  * Consider extra bytes to be an error
  * Decode plain `boolean` and `Boolean` types
  * Throw explicit error if reflection information is missing
  * Throw error if primitive type cannot be explicitly decoded

### CODE REVIEW

File: networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/compression/SnappyCompressor.java [hunk 1]
```diff
@@ -45,31 +47,60 @@ public Bytes compress(final Bytes data) {
...
+      final boolean validateSize = uncompressedPayloadSize.isPresent();
+      if (validateSize && uncompressed.size() < uncompressedPayloadSize.getAsInt()) {
         throw new PayloadSmallerThanExpectedException(
             String.format(
                 "Expected %d bytes but only uncompressed %d bytes",
-                uncompressedPayloadSize, uncompressed.size()));
+                uncompressedPayloadSize.getAsInt(), uncompressed.size()));
       }
-      if (snappyIn.available() > 0) {
+      if (validateSize && snappyIn.available() > 0) {
```
Comment (ajsutton) [inline networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/compression/SnappyCompressor.java]: Shouldn't we be doing this check in all cases? ie even if we don't know what the expected uncompressed size is, we should check that we actually read all bytes from the gossip message.

**Review by ajsutton — DISMISSED** [review]
LGTM. Particularly nice to make SSZ decoding more explicit about failures.

File: networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/compression/SnappyCompressor.java (hunk 1 above)
Comment (mbaxter) [reply networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/compression/SnappyCompressor.java]: It turns out we're supposed to use different compression strategies for gossip vs rpc requests (snappy block-format vs framed-format compression).  So, reverted these changes.

**Review by cemozerr — APPROVED** [review]
LGTM.

**Pull request merged**